labgrid/driver/power/netio_json.py

```diff
@@ -0,0 +1,157 @@
+            outputStates = responseDict['Outputs']
+
+            for outputState in outputStates:
+                if outputState['ID'] == socketID:
+                    state = outputState
+                    break
+        else:
+            raise Exception(f"Cannot SET the power state for socket number: {socketID} for action: {action}."
+                            f" Error code: {response.text}")
+
+        return state
+
+    def get_state(self, socketID):
+        """
+        Get current state of a given socket number as json.
+        """
+        state = None
+        response = requests.get(self.generate_request_url())
+
+        if response.ok:
+            responseDict = json.loads(response.text)
+
+            for outputState in responseDict['Outputs']:
+                if outputState['ID'] == socketID:
+                    state = outputState
+                    break
+        else:
+            raise Exception(f"Cannot get the power state for socket {socketID}. Error code: {response.text}")
+
+        return state
+
+    def convert_socket_id(self, socketID) -> int:
+        """
+        Cast socketID to int.
+
+        raises ValueError
+        """
+        try:
+            socketID = int(socketID)
+        except ValueError as e:
+            raise Exception(f"socketID \"{socketID}\" could not be converted to an integer: {e}!")
+
+        return socketID
+
+    def generate_request_url(self) -> str:
+        """
+        Generate request URL from given params.
+        """
+        requestUrl = 'http://'
+
+        if self.__username and self.__password:
+            requestUrl += f'{self.__username}:{self.__password}@'
+
+        requestUrl += f'{self.__host}:{self.__port}/netio.json'
+
+        return requestUrl
+
+    def get_request_json_data(self, socketID, action) -> str:
+        data = f'{{"Outputs":[{{"ID":{socketID},"Action":{self.convert_action(action)}}}]}}'
+        return json.loads(data)
+
+    def to_str(self, toConvert) -> str:
+        if toConvert == 1 or toConvert is True:
+            toConvert = "1"
+        elif toConvert == 0 or toConvert is False:
+            toConvert = "0"
+        return toConvert
+
+    def convert_action(self, action) -> str:
+        action = self.to_str(action)
+
+        try:
+            assert 0 <= int(action) <= 4
+            return action
+        except AssertionError as ae:
+            raise Exception(f"{ae}\n Action \"{action}\" seems not right, should be between -> '0'-'4'!")
+
+
+def power_set(host, port, index, action):
+    """
+    Set netio-state at index(socketID)
+
+    - host: netio-device adress
+    - port: standard is 80
+    - index: depends on netio-device 1-n (n is the size of netio)
+    - action:
+        - 0: Turn OFF,
+        - 1: Turn ON,
+        - 2: Short OFF delay (restart),
+        - 3: Short ON delay,
+        - 4: Toggle (invert the state)
+    """
+    netio = __NetioControl(host, port)
```

Comment:
Is the class indirection necessary? How are `username` and `password` set?

Reply:
The class indirection isn't strictly necessary, sure it would also be possible to implement this classless. 

The json-api supports/expects username and password, even if they're empty.
afaik NetworkPowerDriver doesn't support to set username and password because NetworkPowerPort doesn't support it.
We/I couldn't find a way to set `username` and `password` at `exporter-config.yaml`.
Because of this another PR is on it's way (hopefully PR-ready within 2 weeks).

Reply:
Username and password are prefabricated.
If you want this to be removed I will remove *username* and *password*.
If it's ok for you I would keep it for possible "future-use".

Reply:
No, the class is not necessary, we can adjust that. Just as note, we use the NetioControl class in other projects as well, we just added the Labgrid API functions `power_set` and `power_get` to reuse that code and get it working with Labgrid. 

How should we continue on that ?